### DISCUSSION TITLE
Typing error when trying to access LoadedReference after `findOne( , , { populate: ['author'] })`

### DISCUSSION THREAD

**Discussion by jgburet** [mikro-orm/mikro-orm#1425] [Q&A] [answered]
The TS compiler complains about me trying to use the loaded reference of a relationship.

First, I retrieve some object using a `findOne` call, to which I passed the `populate` option:
```ts
const myBook = await state.em.findOne(
    Book,
    { /* some where clause */ },
    {
        populate: ['author'],
    },
);
```

Cool, it found something.

Then, I'm trying to pass the `author` attribute of `myBook` to a function expecting to receive an `Author`. That's where things get messy :/

```
error TS2345: Argument of type '{ id?: string | undefined; } & Reference<Author> & LoadedReference<Author, Author>' is not assignable to parameter of type 'Author'.
  Type '{ id?: string | undefined; } & Reference<Author> & LoadedReference<Author, Author>' is missing the following properties from type 'Author': firstName, lastName, books, and 3 more.

104                     myBook.author
```

From there, I thought I simply had forgotten about `.get()` on the relation object, so I went with: 
```ts
myBook.author.get()
```
and it stopped complaining.

However, now the code crashes during runtime when it reaches this point because `get` does not exist.

Then, I tried printing `myBook` to understand what I had retrieved from `findOne`:
```json
{
  "id": "ba1d3b84-f511-4e63-9534-3dbc197e6e68",
  "createdAt": "2021-02-07T18:06:34.000Z",
  "updatedAt": "2021-02-07T18:06:34.000Z",
  "deletedAt": null,
  "author": {
    "id": "c6d25999-a686-4d01-827b-85bb15567f5c",
    "createdAt": "2021-02-07T18:06:34.000Z",
    "updatedAt": "2021-02-07T18:06:34.000Z",
    "deletedAt": null,
    "firstName": "John",
    "lastName": "Doe"
  },
}
```
Okay, so either calling `JSON.stringify` made it look nicely and got rid of some ORM magic on the `author`, either it was just that plain object.

Then, i went further with printing stuff:
```ts
typeof myBook.author // object
typeof myBook.author.get // undefined
instanceof Author // true
```

From here, I'm not sure which direction to take use `author` as an `Author`.
Did I miss something?

**Comment by B4nan** [comment, marked as answer]
Hard to help more without seeing your actual code and entity definitions, so just adding few notes that could shed some light:

> `error TS2345: Argument of type '{ id?: string | undefined; } & Reference<Author> & LoadedReference<Author, Author>' is not assignable to parameter of type 'Author'.` 
- this error should be quite obvious, you are trying to put a `Reference<Author>` to a property that expects just `Author`.
- as you are using the (optional) `Reference` wrapper, I expect you fully read the section about it: https://mikro-orm.io/docs/entity-references
- this is a way to make types stricter, relations are wrapped and you are required to use async `load` method that will fetch them if not yet loaded
- the `get` method is there, always - it is added to the prototype. but when I say "there", I mean on the `Reference` wrapper, not on the naked entity
  - so if you are trying to access `Author.get` it will obviously fail, that is always `undefined`
  - and if you are trying to access `Reference<Author>.get`, it will be always set
  - but on the TS level, the method is added only based on the populate hint, so `findOne(Book, 1, ['author'])` will have `book.author.get` method accessible on the TS level, while `findOne(Book, 1)` will not

> Okay, so either calling JSON.stringify made it look nicely and got rid of some ORM magic on the author, either it was just that plain object.

Read about how serialization work, entity is not DTO and vice versa. If you convert entity to JSON, you will end up with DTO. Reference wrapper is purely runtime thing, it is removed when serialized as it has no meaning in the API response. 

https://mikro-orm.io/docs/serializing/

> typeof myBook.author // object
> typeof myBook.author.get // undefined
> instanceof Author // true

This means you have it wrong, the `myBook.author` is either defined as a naked entity, or a reference wrapped. Sounds like you are missing `wrappedReference: true` in the decorator - it is mandatory for references unless you use ts-morph (or you want naked entity there).

```ts
@ManyToOne(() => Author, { wrappedReference: true })
author!: IdentifiedReference<Author>;
```

**Reply by jgburet** [reply]
Thank you for the quick answer!

What I missed was this `wrappedReference: true`. Adding it solved everything.

Those `get`, `load`, etc methods were not present (and were crashing when reached, because I was using them) until I added the flag on the model relations decorators.
